Simplify packaging and simplify the function used to read FASTQ files

paalka opened the pull request:
Sorry for making so many changes in one pull request, this is far from ideal, so feel very free to ask questions about the changes, or making me split it into smaller pull requests  :)

The changes can be quickly summarized as an attempt to simplify some of the functionality and hopefully make it easier to distribute.

The project has been entirely restructured in order to make it possible to use Maven instead of a Makefile, as this make it easier to handle dependencies and build JARs in consistent way.
I've also made it so that the libbwa file is packaged and loaded from inside the JAR, as this makes it easier to use.

All the code has also been reformatted to follow Google's Java styleguide, since this IMO makes the code easier to read.

Previously, when reading FASTQ files, some of the reads would not be included (see ex. http://www.bioplanet.com/gcat/reports/7770-ecfkcezhcs/alignment/100bp-pe-small-indel/sparkbwa-mem/compare-8009-uyddaboumb-8003-oqbwsapgew -- where SparkBWATwo is this version).

The reason for this is most likely because Hadoop would sometimes split records between nodes. This has been improved by using Spark directly for handling the reads instead of Hadoop. This also makes it easier to use SparkBWA with other filesystems (such as just NFS).  Some of the reads are still missing, but it is at least less missing reads.

paalka committed:
Ignore common Java, VIM, C, and Intelij related tmp files.

paalka committed:
Restructure the project so that it is easier to use with other projects, and so
that it follows the standard Java way of structuring projects.

paalka committed:
Use Maven to build the JAR file instead of a Makefile, as this makes it easier
to import into other projects.

paalka committed:
Reformat the entire project so that it follows Google's Java styleguide.

paalka committed:
Remove two classes which are no longer used.

paalka committed:
Remove a function which does not really work as intended.

paalka committed:
Ensure that the output dir is created if it does not exist.

paalka committed:
Attempt to make it easier to copy the result to non-hdfs locations.

paalka committed:
Ensure that Java 1.8 is used when compiling the classes, as this makes it
easier to perform certain operations.

paalka committed:
Simplify the functions used to parse the FASTQ files.

The FASTQ records are now parsed directly using Spark, instead of using
custom Hadoop record classes. When using Hadoop record classes the input file
system was limited to HDFS, which is not always ideal.

paalka committed:
Move the BWA files to a folder which better describes what it is used as.

paalka committed:
Ensure that the new lib folder is used when attempting to compile BWA.

paalka committed:
Use the native-utils library for packaging the bwa lib inside the jar.

paalka committed:
Do not track libbwa.so, as it may change often.

paalka committed:
Do not track the reduced dependency POM generated by maven.

paalka committed:
Ensure that the FASTQ records are built in the correct order, as the iterator
may not always have placed the elements of the record in the correct order
(i.e. by line number).

paalka committed:
Make sure that the resources folder exists before copying the shared lib to it.

paalka committed:
Make an empty constructor, which is useful when integrating the program into
other applications.

paalka committed:
Remove a hard coded reference to a tmp folder, which was added by mistake.

paalka committed:
No longer track object files and static libraries, as this are generated when
building BWA.

paalka committed:
Remove all references to bwa.zip, as bwa is now packaged inside the generated
JAR file.

paalka committed:
Use the same naming scheme for all functions.

jmabuin commented:
Thanks, I will check the code before accept it. As I said before you have implemented some of my ideas, such as use maven and include libbwa.so inside the jar, and improve a lot of points, thanks!

paalka committed:
Remove a function which is no longer needed, as libbwa is packaged inside the
JAR.

paalka committed:
Simplify the mappers used to group the lines in a record.

Instead of using regexes to attempt to determine the type of the line, just
use the line number to determine the type.

paalka committed:
Make sure that the JAR checksums are not included inside the shaded JAR, as
this causes errors.

paalka committed:
Simplify the logic used to get the FASTQ record number.
